File: go-backend/internal/format/speed.go
package format

import "fmt"

func SpeedToMbps(bytesPerSecond float64) string {
	if bytesPerSecond <= 0 {
		return "0 Kbps"
	}

	kbps := bytesPerSecond * 8 / 1000
	if kbps < 1000 {
		return fmt.Sprintf("%.2f Kbps", kbps)
	}

	mbps := kbps / 1000
	if mbps < 1000 {
		return fmt.Sprintf("%.2f Mbps", mbps)
	}

	return fmt.Sprintf("%.2f Gbps", mbps/1000)
}

File: vite-frontend/src/utils/speed.ts
export const formatSpeedToMbps = (bytesPerSecond: number): string => {
  if (!Number.isFinite(bytesPerSecond) || bytesPerSecond <= 0) {
    return "0 Kbps";
  }

  const bitsPerSecond = bytesPerSecond * 8;
  const kbps = bitsPerSecond / 1000;

  if (kbps < 1000) {
    return `${formatSpeedNumber(kbps)} Kbps`;
  }

  const mbps = kbps / 1000;

  if (mbps < 1000) {
    return `${formatSpeedNumber(mbps)} Mbps`;
  }

  return `${formatSpeedNumber(mbps / 1000)} Gbps`;
};

export const formatMegaBytesToMbps = (megaBytesPerSecond: number): string =>
  formatSpeedToMbps(megaBytesPerSecond * 1024 * 1024);

const formatSpeedNumber = (value: number): string => {
  if (!Number.isFinite(value) || value <= 0) {
    return "0";
  }

  return value.toFixed(2);
};
